core/src/main/java/hudson/model/Actionable.java

```diff
@@ -134,17 +141,99 @@ public void addAction(Action a) {
      * @param a an action to add/replace
      * @since 1.548
      */
-    public void replaceAction(Action a) {
+    @SuppressWarnings({"ConstantConditions", "deprecation"})
+    @SuppressFBWarnings("RCN_REDUNDANT_NULLCHECK_OF_NONNULL_VALUE")
+    public void replaceAction(@Nonnull Action a) {
+        if (a == null) {
+            throw new IllegalArgumentException("Action must be non-null");
+        }
         // CopyOnWriteArrayList does not support Iterator.remove, so need to do it this way:
         List<Action> old = new ArrayList<Action>(1);
         List<Action> current = getActions();
+        boolean found = false;
         for (Action a2 : current) {
-            if (a2.getClass() == a.getClass()) {
+            if (!found && a.equals(a2)) {
+                found = true;
+            } else  if (a2.getClass() == a.getClass()) {
                 old.add(a2);
             }
         }
         current.removeAll(old);
-        addAction(a);
+        if (!found) {
+            addAction(a);
+        }
+    }
+
+    /**
+     * Remove an action.
```

Comment:
Should clarify that `getAllActions()` could still contain it—just not `getActions()`.